cli/src/program.rs

```diff
@@ -1129,22 +1289,173 @@ fn process_program_deploy(
             None,
         )?;
     }
-    if result.is_err() && buffer_signer_index.is_none() {
-        report_ephemeral_mnemonic(words, mnemonic);
+    if result.is_err() && !buffer_provided {
+        // We might have deployed "temporary" buffer but failed to deploy our program from this
+        // buffer, reporting this to the user - so he can retry deploying re-using same buffer.
+        report_ephemeral_mnemonic(buffer_words, buffer_mnemonic);
     }
     result
 }
 
+/// Upgrade existing program using upgradeable loader
+#[allow(clippy::too_many_arguments)]
+fn process_program_upgrade(
+    rpc_client: Arc<RpcClient>,
+    config: &CliConfig,
+    fee_payer_signer_index: SignerIndex,
+    program_signer_index: SignerIndex,
+    buffer_signer_index: SignerIndex,
+    upgrade_authority_signer_index: SignerIndex,
+    is_final: bool,
+    skip_fee_check: bool,
+    sign_only: bool,
+    blockhash_query: &BlockhashQuery,
+) -> ProcessResult {
+    let fee_payer_signer = config.signers[fee_payer_signer_index];
+    let upgrade_authority_signer = config.signers[upgrade_authority_signer_index];
+    let buffer_pubkey = config.signers[buffer_signer_index].pubkey();
+    let program_pubkey = config.signers[program_signer_index].pubkey();
+
+    let blockhash = blockhash_query.get_blockhash(&rpc_client, config.commitment)?;
+    let final_message = Message::new_with_blockhash(
+        &[bpf_loader_upgradeable::upgrade(
+            &program_pubkey,
+            &buffer_pubkey,
+            &upgrade_authority_signer.pubkey(),
+            &fee_payer_signer.pubkey(),
+        )],
+        Some(&fee_payer_signer.pubkey()),
+        &blockhash,
+    );
+
+    if sign_only {
+        let mut tx = Transaction::new_unsigned(final_message);
+        let signers = &[fee_payer_signer, upgrade_authority_signer];
+        // Using try_partial_sign here because fee_payer_signer might not be the fee payer we
+        // end up using for this transaction (it might be NullSigner).
+        tx.try_partial_sign(signers, blockhash)?;
+        return_signers_with_config(
+            &tx,
+            &config.output_format,
+            &ReturnSignersConfig {
+                dump_transaction_message: true, // print out resulting signature
```

Comment:
My reasoning is: do we need this to ever be `false` ? And I can't find a plausible "yes" answer - hence the simplest way is to just have `true` hardcoded here,

what's the use-case for not printing out signature ?

Reply:
This controls dumping just the transaction message as base-64. The transaction message does not contain any signatures, just the compiled instructions. The signatures will always be output separately.

In my experience, I've rarely dumped the transaction message because I just wanted the signature. And since the arg is already added, it makes more sense to actually use it here